### PR TITLE
WIP: Support kubeflow.org/pytorchjob

### PR DESCRIPTION
**Pull request by tenzen-y**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Support kubeflow.org/pytorchjob.

#### Which issue(s) this PR fixes:

Part-of #297

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+func GetWorkloadNameForPyTorchJob(jobName string) string {
+	return jobframework.GetWorkloadNameForOwnerWithGVK(jobName, gvk)
+}
+
+type PyTorchJob struct {
+	*kftraining.PyTorchJob
+	*kubeflowjob.KubeflowJob
+}
```
Comment (trasc) [inline pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:56]: If you make `Object()` mart of the `KFJobControl` you could remove the `kftraining.PyTorchJob` member,  and potentially change this  to :

```suggestion
type PyTorchJob kubeflowjob.KubeflowJob
```

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+func (c *JobControl) ReplicaSpecs() map[kftraining.ReplicaType]*kftraining.ReplicaSpec {
+	return c.Spec.PyTorchReplicaSpecs
+}
+
+func (c *JobControl) RunPolicy() *kftraining.RunPolicy {
+	return &c.Spec.RunPolicy
+}
+
+func (c *JobControl) OrderedReplicaTypes(replicaSpecs map[kftraining.ReplicaType]*kftraining.ReplicaSpec) []kftraining.ReplicaType {
+	var result []kftraining.ReplicaType
```
Comment (trasc) [inline pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:108]:
```suggestion
	result := make([]kftraining.ReplicaType, 0, 2)
```

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go
Comment (trasc) [inline pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:1]: Missing header

File: pkg/controller/jobs/kubeflow/jobs/jobs.go
Comment (trasc) [inline pkg/controller/jobs/kubeflow/jobs/jobs.go:1]: Missing header

also I think we could just import 
`"sigs.k8s.io/kueue/pkg/controller/jobs/kubeflow/jobs/pytorchjob"`
in 
`pkg/controller/jobs/jobs.go`

File: pkg/controller/jobs/kubeflow/kubeflowjob/interface.go
Comment (trasc) [inline pkg/controller/jobs/kubeflow/kubeflowjob/interface.go:1]: missing header

File: pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob.go
Comment (trasc) [inline pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob.go:1]: missing header

**Review by trasc — COMMENTED** [review]
The generic approach around `pkg/controller/jobs/kubeflow/kubeflowjob` makes the implementation a bit hard to follow. Are we planing to support other similar kubeflow jobs in the near future?

**Comment by tenzen-y** [comment]
> The generic approach around pkg/controller/jobs/kubeflow/kubeflowjob makes the implementation a bit hard to follow. Are we planing to support other similar kubeflow jobs in the near future?

@trasc Yes. We will support TFJob, MXJob, XGboostJob, and PaddleJob.
cc: @alculquicondor

File: config/components/rbac/role.yaml [hunk 3]
```diff
@@ -122,6 +122,23 @@ rules:
...
+  - patch
+  - update
+  - watch
+- apiGroups:
+  - kubeflow.org
+  resources:
+  - pytorchjobs/status
+  verbs:
+  - get
+  - update
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply config/components/rbac/role.yaml:141]: It's a good point. Thanks.

File: go.mod [hunk 4]
```diff
@@ -83,3 +85,5 @@ require (
 	sigs.k8s.io/json v0.0.0-20221116044647-bc3834ca7abd // indirect
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
+
+replace github.com/kubeflow/training-operator v1.6.0 => github.com/tenzen-y/training-operator v1.3.0-rc.1.0.20230717233919-1ed3e8e55322
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply go.mod:89]: Yes, I will do `b`.

File: pkg/controller/jobs/kubeflow/jobs/jobs.go
Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/jobs.go:1]: In the future, we will add more framework support. So, I would like to put the linking for kubeflow here.

```go
// Reference the job framework integration packages to ensure linking.
import (
	_ "sigs.k8s.io/kueue/pkg/controller/jobs/kubeflow/jobs/pytorchjob"
	_ "sigs.k8s.io/kueue/pkg/controller/jobs/kubeflow/jobs/tfjob"
	_ "sigs.k8s.io/kueue/pkg/controller/jobs/kubeflow/jobs/paddlejob"
	_ "sigs.k8s.io/kueue/pkg/controller/jobs/kubeflow/jobs/xgboostjob"
	_ "sigs.k8s.io/kueue/pkg/controller/jobs/kubeflow/jobs/mxjob"
)
```

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go (hunk 1 above)
Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:56]: It makes sense.

Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:56]: For Composition, we need to use an embedded struct. So I will modify this in the following:

```go
type PyTorchJob struct {
	*kubeflowjob.KubeflowJob
}
```

Comment (trasc) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:56]: I think it should work the same way, it is used like this to "make" the generic job out of all integrations we currently have.

Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:56]: @trasc  Does that mean we should use the type alias instead of the embedded struct?

If doing so, it means that we remove `kubeflowjob.KubeflowJob`, and then in all kubeflow integrations (TFJob, PaddleJob, ...), we declare the same processes such as `RunWithPodSetsInfo`, `RestorePodSetsInfo`  and so on.

Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:56]: I'd like to use the embedded struct instead of the type alias because I'd like to use method override and then avoid redeclaration of the same processes.

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go (hunk 2 above)
Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go]: Done.

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go
Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:1]: Done.

File: pkg/controller/jobs/kubeflow/jobs/jobs.go
Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/jobs.go:1]: > Missing header

Done.

File: pkg/controller/jobs/kubeflow/kubeflowjob/interface.go
Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/kubeflowjob/interface.go:1]: Done.

File: pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob.go
Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob.go:1]: Done.

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go (hunk 1 above)
Comment (trasc) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go]: The embed approach comes with an extra  allocation every time is passed by value.  Method overwrite should work the same way.  (this is not type alias)

Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go]: Hm. My understanding might not be correct. Could you clarify more about `If you make Object() mart of the KFJobControl`?

Comment (trasc) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go]: If Object is part of `KFJobControl`, then you can  change 

```diff
func (j *KubeflowJob) Object() client.Object {
	return j.KFJobControl.Object()
}
```

Then you don't need the `*kftraining.PyTorchJob` memeber. 

Then you can simplify the struct, and also make it a type "redefinition"  `type PyTorchJob kubeflowjob.KubeflowJob`.

Comment (trasc) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go]: thinking about it , you could do the same for `GetGVK() ` and might be able to make `KubeflowJob` generic and

```diff
var NewReconciler = jobframework.NewGenericReconciler(func() jobframework.GenericJob {
-	pytorchJobObj := &kftraining.PyTorchJob{}
-	return &PyTorchJob{PyTorchJob: pytorchJobObj, KubeflowJob: kubeflowjob.NewKubeflowJob((*JobControl)(pytorchJobObj))}
+	return &kftraining.PyTorchJob{}
}, nil)

...

-type PyTorchJob struct {
-	*kftraining.PyTorchJob
-	*kubeflowjob.KubeflowJob
-}
+ type PyTorchJob  KubeflowJob[PyTorchJob]
```

Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go]: > If Object is part of KFJobControl, then you can change
> 
> func (j *KubeflowJob) Object() client.Object {
> -	return nil
> +	return j.KFJobControl.Object()
> }
> Then you don't need the *kftraining.PyTorchJob memeber.
> 
> Then you can simplify the struct, and also make it a type "redefinition" type PyTorchJob kubeflowjob.KubeflowJob.

Thanks for the clarifications.
If doing so, I think `type PyTorchJob kubeflowjob.KubeflowJob` doesn't implement the `GenericJob` interface, right?

Although `type JobControl kftraining.PyTorchJob` implement the `GenericJob`.

Comment (trasc) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go]: Tooo many interfaces :)... ,  

To keep it simple I think `kubeflowjob.KubeflowJob` should have all the common code for kubeflow(kftraining)) Jobs. and use `KFJobControl` to get it's job  (eg. `kftraining.PyTorchJob`) specific bits.

To avoid confusion,  should be a named member, don't embed.

Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go]: Uhm. Actually, I adopted the same pattern you suggested before submitting this PR. Then, I switched implementation to now one since this one is more reusable.

However, I will roll back my implementation to your suggested one since if @trasc,  who has much understanding of the kueue, is confused by this implementation, then almost everyone will be confused.

**Comment by alculquicondor** [comment]
I'll get back to this after @trasc gives an LGTM.
I think he's out this week.

**Comment by tenzen-y** [comment]
Sorry, I've been away from this PR for a while.
I will move this forward in the next week since I'm busy now due to Kubeflow code freeze.

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go (hunk 1 above)
Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go]: Done.

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply go.mod]: Done.

File: config/components/rbac/role.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply config/components/rbac/role.yaml:141]: Done.

**Review by trasc — COMMENTED** [review]
/lgtm

File: Makefile [hunk 5]
```diff
@@ -153,7 +153,7 @@ test: generate gotestsum ## Run tests.
 	$(GOTESTSUM) --junitfile $(ARTIFACTS)/junit.xml -- $(GO_TEST_FLAGS) $(shell $(GO_CMD) list ./... | grep -v '/test/') -coverpkg=./... -coverprofile $(ARTIFACTS)/cover.out
 
 .PHONY: test-integration
-test-integration: manifests generate envtest ginkgo mpi-operator-crd ray-operator-crd jobset-operator-crd ## Run tests.
+test-integration: manifests generate envtest ginkgo mpi-operator-crd ray-operator-crd jobset-operator-crd training-operator-crd ## Run tests.
```
Comment (alculquicondor) [inline Makefile:156]: "training-operator" sounds a bit generic, in a way that can be confusing for future readers.
Maybe "tf-training-operator"?

File: Makefile [hunk 6]
```diff
@@ -320,6 +320,12 @@ mpi-operator-crd:
 	mkdir -p $(PROJECT_DIR)/dep-crds/mpi-operator/
 	cp -f $(MPIROOT)/manifests/base/* $(PROJECT_DIR)/dep-crds/mpi-operator/
 
+TRAININGROOT = $(shell $(GO_CMD) list -m -f "{{.Dir}}" github.com/kubeflow/training-operator)
```
Comment (alculquicondor) [inline Makefile:323]:
```suggestion
TFTRAININGROOT = $(shell $(GO_CMD) list -m -f "{{.Dir}}" github.com/kubeflow/training-operator)
```

File: pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_controller.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+func (j *KubeflowJob) Object() client.Object {
+	return j.KFJobControl.Object()
+}
+
+func (j *KubeflowJob) IsSuspended() bool {
+	return j.KFJobControl.RunPolicy().Suspend != nil && *j.KFJobControl.RunPolicy().Suspend
+}
+
+func (j *KubeflowJob) Suspend() {
```
Comment (alculquicondor) [inline pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_controller.go:50]: Can RunPolicy be nil? Should we add it here if it's missing?

File: pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_controller.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+	j.KFJobControl.RunPolicy().Suspend = pointer.Bool(true)
+}
+
+func (j *KubeflowJob) RunWithPodSetsInfo(podSetInfos []jobframework.PodSetInfo) error {
+	j.KFJobControl.RunPolicy().Suspend = pointer.Bool(false)
```
Comment (alculquicondor) [inline pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_controller.go:55]: Same note about nil RunPolicy

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+func (j *JobControl) ReplicaSpecs() map[kftraining.ReplicaType]*kftraining.ReplicaSpec {
+	return j.Spec.PyTorchReplicaSpecs
+}
+
+func (j *JobControl) JobStatus() kftraining.JobStatus {
+	return j.Status
+}
+
+func (j *JobControl) OrderedReplicaTypes(replicaSpecs map[kftraining.ReplicaType]*kftraining.ReplicaSpec) []kftraining.ReplicaType {
```
Comment (alculquicondor) [inline pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:88]: What if you make this function just return the list of ReplicaTypes? Wouldn't the KFJob be able to obtain the rest?

File: Makefile (hunk 5 above)
Comment (tenzen-y) [reply Makefile:156]: How about "kf-training-operator"?

Comment (alculquicondor) [reply Makefile:156]: oops, that's what I wanted to say, sorry

File: pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_controller.go (hunk 7 above)
Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_controller.go:50]: RunPolicy isn't a pointer. So we wouldn't be concerned about missing Runpolicy.

https://github.com/kubeflow/training-operator/blob/48dbbf0a8e90e52c55ec05d0f689fcbf83c6b441/pkg/apis/kubeflow.org/v1/pytorch_types.go#L76-L80

@alculquicondor WDYT?

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go (hunk 9 above)
Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:88]: IIUC, it wouldn't work since every kubeflow framework has different replicaRoles.
For example:

TFJob: https://github.com/kubeflow/training-operator/blob/48dbbf0a8e90e52c55ec05d0f689fcbf83c6b441/pkg/apis/kubeflow.org/v1/tensorflow_types.go#L104-L123
PyTorchJob: https://github.com/kubeflow/training-operator/blob/48dbbf0a8e90e52c55ec05d0f689fcbf83c6b441/pkg/apis/kubeflow.org/v1/pytorch_types.go#L40-L43
MXJob: https://github.com/kubeflow/training-operator/blob/48dbbf0a8e90e52c55ec05d0f689fcbf83c6b441/pkg/apis/kubeflow.org/v1/mxnet_types.go#L39-L58

Comment (alculquicondor) [reply pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go:88]: Nvm. This function is doing exactly what I said. I read it wrong.

File: Makefile (hunk 5 above)
Comment (tenzen-y) [reply Makefile]: Done.

File: Makefile (hunk 6 above)
Comment (tenzen-y) [reply Makefile]: Done.

**Comment by tenzen-y** [comment]
The kueue-controller-manager seems suddenly can not access the kube-apiserver:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/995/pull-kueue-test-e2e-main-1-24/1691153180097777664/artifacts/run-test-e2e-1.24.15/kueue-controller-manager.log
/retest

**Comment by tenzen-y** [comment]
Squashed.

File: pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_controller.go (hunk 7 above)
Comment (tenzen-y) [reply pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_controller.go:50]: @alculquicondor friendly ping :)

Comment (alculquicondor) [reply pkg/controller/jobs/kubeflow/kubeflowjob/kubeflowjob_controller.go:50]: sounds good :)

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
Support kubeflow.org/pytorchjob
```
/hold cancel
